### PR TITLE
feat(remap transform): add error output stream

### PR DESCRIPTION
**Pull request by lukesteensen**
The details of the UI here are a little iffy, but this seemed like the smallest change that was backwards-compatible and didn't introduce a third option about what to do with the event when something goes wrong. I fully expect some discussion and resulting changes here, but it seemed worthwhile to put this up as a starting point.

### CODE REVIEW

File: src/transforms/remap.rs [hunk 1]
```diff
@@ -169,6 +173,8 @@ impl FunctionTransform for Remap {
 
                 if !self.drop_on_error {
                     output.push(original_event.expect("event will be set"))
+                } else {
+                    err_output.push(original_event.expect("event will be set"))
                 }
```
Comment (lukesteensen) [inline src/transforms/remap.rs:178]: Piggybacking on this also feels weird, but it is rather simple and has the one large advantage of being both backwards compatible and not introducing any new knobs. Very open to any other suggestions.

**Comment by hhromic** [comment]
This is a very interesting change, it would make it possible to do this kind of pipelines?
```yaml
transforms:
  mytr1:
    type: remap
    inputs:
      - logs-source
    drop_on_errors: true
    source: |-
      parse_csv!(.message)
  mytr2:
    type: remap
    inputs:
      - logs-source
    drop_on_errors: true
    source: |-
      parse_json!(.message)
  log-errors:
    type: remap
    inputs:
      - mytr1.errors
      - mytr2.errors
    source: |-
      log("invalid message: " + string!(.message), level: "error")
```

File: src/transforms/remap.rs (hunk 1 above)
Comment (JeanMertz) [reply src/transforms/remap.rs:178]: I'm not opposed to this solution, as I think we'll be tweaking `drop_on_error` and `drop_on_abort` in the future regardless (see https://github.com/vectordotdev/vector/issues/7496).

The one problem I see with this is that you can't use VRL's `abort` keyword in this case to reroute. Perhaps we should also add an `aborts` named output for aborted programs?

The other downside to this, is that there's like a (minor?) performance drop if you just want to drop events and never read from the errors/aborts channels. It would be nice if we could somehow short-circuit sending events over those channels if there's no need to, without having the user have to specify that through a config flag (e.g. we'd need the topology to somehow expose a list of components listening on those specific inputs).

But, that last point isn't blocking to me, and it seems that `aborts` is an easy one to add, if we agree on that.

**Comment by lukesteensen** [comment]
@hhromic

Exactly! This allows for sending events that fail to process down a different pipeline without any manual tricks to catch errors, tag, fanout, filter, etc. Over time we can extend it to more situations where explicit branching would be desirable.

**Comment by JeanMertz** [comment]
> * Are we ok with piggybacking on `drop_on_error` like this? It's not the clearest, but it's very simple and backwards-compatible. Especially if we're going to revisit these options in the near future, I would lean towards yes.

I'm okay with this 👍

> * Do we want to do anything with `abort` right now? The options would be to write them to the same output (trivial) or to do the work to support a third output (less trivial). I'd lean towards deferring for now.

Given the added complexity, and the fact that we'll likely be changing the `drop_on_*` pattern in this transform at some point, I'm also okay with deferring for now.

> * Implementation-wise, do we have any reservations about `impl<T> FunctionTransform for T where T: FallibleFunctionTransform`?

Could we add some kind of `cfg(not(test))` condition to a `panic` inside that implementation that triggers when `err_buf` is non-empty? That is, in non-test mode, when using this for a transform that's supposed to be fallible, we want to be sure to catch the invalid usage.

**Comment by lukesteensen** [comment]
> is the metadata object included in the messages going to sinks, or not by default? (Wasn't included in the `md` file example)

Good catch, updated the example with the new output.

**Comment by JeanMertz** [comment]
One thing that came up during our check-in is that we might consider not making this the "error output stream", but the "dropped messages stream". That is, messages can be dropped because of errors, or because of an explicit `abort` statement.

With the added metadata that includes the reason a message was dropped, we could merge both into the "dropped" stream, and people could inspect the metadata to understand why the message was dropped.

**Comment by binarylogic** [comment]
I agree with @JeanMertz's naming. The main reason we want to enable this is to ensure data is not lost.

**Comment by lukesteensen** [comment]
Sounds good, just pushed an update to call it `dropped` and include `abort` as well as normal errors.

File: website/content/en/highlights/2021-11-04-remap-dropped-output.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  type: "new feature"
+  domains: ["remap"]
+---
+
+Vector 0.18 introduces a new `dropped` output to the `remap` transform. This
+can be used to route events that fail processing down a separate pipeline.
+
+When either of `drop_on_error` or `drop_on_abort` is set to `true`, events that
+are dropped from the primary output stream due to either errors or aborts are
+written instead to a separate output called `dropped`.
```
Comment (JeanMertz) [inline website/content/en/highlights/2021-11-04-remap-dropped-output.md:19]: It's worth pointing out that none of the mutations before the error/drop-statement are preserved.

The example below does show that, but it's worth pointing out in a single sentence for people that skip the example.

File: website/content/en/highlights/2021-11-04-remap-dropped-output.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    'invalid message',
+  ]
+
+[transforms.remap]
+  type = "remap"
+  inputs = ["in"]
+  drop_on_error = true
+  drop_on_abort = true
+  source = """
+    . |= object!(parse_json!(.message))
```
Comment (JeanMertz) [inline website/content/en/highlights/2021-11-04-remap-dropped-output.md:46]: note: I've seen this pattern being used a lot, I'm tracking this in https://github.com/vectordotdev/vector/issues/9927.

File: website/content/en/highlights/2021-11-04-remap-dropped-output.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+{"message":"{ \"message\": \"valid message\", \"foo\": \"baz\"}","metadata":{"component":"remap","error":"aborted"},"timestamp":"2021-11-05T00:42:10.945183635Z"}
+{"message":"invalid message","metadata":{"component":"remap","error":"function call error for \"object\" at (9:39): function call error for \"parse_json\" at (17:38): unable to parse json: expected value at line 1 column 1"},"timestamp":"2021-11-05T00:42:11.944980725Z"}
+{"message":"invalid message","metadata":{"component":"remap","error":"function call error for \"object\" at (9:39): function call error for \"parse_json\" at (17:38): unable to parse json: expected value at line 1 column 1"},"timestamp":"2021-11-05T00:42:12.944970623Z"}
+{"foo":"bar","message":"valid message","processed":true,"timestamp":"2021-11-05T00:42:13.945360616Z"}
+```
+
+All of the events that were valid JSON were processed and output as JSON via the
+`out` console sink, while those that failed are written out in plain text via the
+`errors_out` console sink.
```
Comment (JeanMertz) [inline website/content/en/highlights/2021-11-04-remap-dropped-output.md:82]: > while those that failed are written out in plain text via the `errors_out` console sink.

This is only the case in this example because we use the unparsed `generator` events, I think it's best to mention that they are written out as the original event shape before the remap transform ran.

File: website/content/en/highlights/2021-11-04-remap-dropped-output.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    }
+    .processed = true
+  """
+
+[sinks.out]
+  type = "console"
+  inputs = ["remap"]
+  encoding.codec = "json"
+
+[sinks.errors_out]
```
Comment (JeanMertz) [inline website/content/en/highlights/2021-11-04-remap-dropped-output.md:58]:
```suggestion
[sinks.bar_out]
```

nit: to clarify that this isn't related to "errors" per sé, it can be used for whatever reason people want.

File: website/cue/reference/components/transforms/remap.cue [hunk 6]
```diff
@@ -8,6 +8,13 @@ components: transforms: "remap": {
 		[Vector Remap Language](\(urls.vrl_reference)) (VRL), an expression-oriented language designed for processing
 		observability data (logs and metrics) in a safe and performant manner.
 
+		This transform also implements an additional `errors` output. When the
+		`drop_on_error` configuration value is set to `true`, events that result in
+		runtime errors will be dropped from the default output stream and sent to
+		the `errors` output instead. For a transform component named `foo`, this
+		error output can be accessed by specifying `foo.errors` as the input to
+		another component.
```
Comment (JeanMertz) [inline website/cue/reference/components/transforms/remap.cue:16]: This needs to be rewritten to also mention `drop_on_abort`, and rename `errors` output to `dropped`.

File: website/cue/reference/components/transforms/remap.cue [hunk 7]
```diff
@@ -89,7 +96,9 @@ components: transforms: "remap": {
 			common:   false
 			required: false
 			description: """
-				Drop the event if the VRL program returns an error at runtime.
+				Drop the event from the primary output stream if the VRL program returns
+				an error at runtime. These events will instead be written to the
+				`errors` output.
```
Comment (JeanMertz) [inline website/cue/reference/components/transforms/remap.cue:101]: Same here: re: runtime error or explicit abort, and stream is named `dropped` now.

**Review by JeanMertz — APPROVED** [review]
A few lingering points, but otherwise LGTM.

File: lib/vector-core/src/transform/config.rs [hunk 8]
```diff
@@ -15,19 +15,17 @@ pub enum ExpandType {
     Serial,
 }
 
-#[cfg(feature = "vrl")]
 #[derive(Debug, Default)]
 pub struct TransformContext {
+    // This is optional because currently there are a lot of places we use `TransformContext` that
+    // may not have the relevant data available (e.g. tests). In the future it'd be nice to make it
+    // required somehow.
+    pub key: Option<ComponentKey>,
```
Comment (jszwedko) [inline lib/vector-core/src/transform/config.rs:23]: Agreed, it'd be nice to remove the option here in the future.

File: src/transforms/remap.rs [hunk 9]
```diff
@@ -152,13 +156,25 @@ impl FunctionTransform for Remap {
...
+                    let mut erred_event = original_event.expect("event will be set");
+                    if let Event::Log(ref mut log) = &mut erred_event {
+                        log.insert(
+                            log_schema().metadata_key(),
+                            serde_json::json!({
+                                "error": error.to_string(),
+                                "component": self.component_key,
+                            }),
+                        );
+                    }
```
Comment (jszwedko) [inline src/transforms/remap.rs:176]: I probably missed this earlier, but is there a reason we are only doing this for logs and not also metrics which `remap` can be used for?

File: src/transforms/remap.rs [hunk 10]
```diff
@@ -169,6 +185,18 @@ impl FunctionTransform for Remap {
 
                 if !self.drop_on_error {
                     output.push(original_event.expect("event will be set"))
+                } else {
+                    let mut erred_event = original_event.expect("event will be set");
+                    if let Event::Log(ref mut log) = &mut erred_event {
+                        log.insert(
+                            log_schema().metadata_key(),
+                            serde_json::json!({
+                                "error": error.to_string(),
```
[the remark this answers is not in the document]
Comment (jszwedko) [reply src/transforms/remap.rs:194]: I like this, I could see us adding some sort of notion of "error codes" in the future too.

File: website/content/en/highlights/2021-11-04-remap-dropped-output.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+pr_numbers: [9169]
+release: "0.18.0"
+hide_on_release_notes: false
+badges:
+  type: "new feature"
+  domains: ["remap"]
+---
+
+Vector 0.18 introduces a new `dropped` output to the `remap` transform. This
+can be used to route events that fail processing down a separate pipeline.
```
Comment (jszwedko) [inline website/content/en/highlights/2021-11-04-remap-dropped-output.md:15]: If we expect to only support logs, we should make that clearer in this highlight.

Also, thanks for proactively writing a highlight for this! I usually end up writing these during the release process so this is welcome.

File: website/content/en/highlights/2021-11-04-remap-dropped-output.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+  inputs = ["remap.dropped"]
+  encoding.codec = "json"
+```
+
+You would expect to see output like the following:
+
+```json
+{"foo":"bar","message":"valid message","processed":true,"timestamp":"2021-11-05T00:42:03.945157398Z"}
+{"foo":"bar","message":"valid message","processed":true,"timestamp":"2021-11-05T00:42:04.945155276Z"}
+{"message":"{ \"message\": \"valid message\", \"foo\": \"baz\"}","metadata":{"component":"remap","error":"aborted"},"timestamp":"2021-11-05T00:42:05.945588208Z"}
```
Comment (jszwedko) [inline website/content/en/highlights/2021-11-04-remap-dropped-output.md:69]: I think users might be used to seeing the same component tags that we have in other places (like metrics). Could we add the other tags here (`component_id`, `component_kind`, and `component_type`?). This would make:

```json
{"message":"{ \"message\": \"valid message\", \"foo\": \"baz\"}","metadata":{"component":"remap","error":"aborted"},"timestamp":"2021-11-05T00:42:05.945588208Z"}
```

Look like:

```json
{"message":"{ \"message\": \"valid message\", \"foo\": \"baz\"}","metadata":{"component_id":"remap","component_kind":"transform","component_type":"remap","error":"aborted"},"timestamp":"2021-11-05T00:42:05.945588208Z"}
```

I think this parity with metrics would be nice.

File: lib/vector-core/src/config/log_schema.rs [hunk 13]
```diff
@@ -53,6 +48,8 @@ pub struct LogSchema {
     host_key: String,
     #[serde(default = "LogSchema::default_source_type_key")]
     source_type_key: String,
+    #[serde(default = "LogSchema::default_metadata_key")]
+    metadata_key: String,
```
Comment (jszwedko) [inline lib/vector-core/src/config/log_schema.rs:52]: I think we need to add this new schema key to the docs still.

**Review by jszwedko — COMMENTED** [review]
Nice! Thanks for the refactor to add metadata. I agree with Jean's comments below and left a few of my own.

File: src/transforms/remap.rs (hunk 9 above)
Comment (lukesteensen) [reply src/transforms/remap.rs]: This type of data (e.g. error messages, etc) didn't really strike me as making sense to shove into metrics tags at the time, but I don't really have a strong opinion on that. If we think they'd be useful I'm happy to add some equivalent to metrics as well.

Comment (jszwedko) [reply src/transforms/remap.rs]: Ah, yeah, that is a wrinkle if we allow routing of failed metrics. 🤔  Overall routing of failed metrics is probably of a less common use-case. I could see supporting it just for parity though, adding the metadata as metrics tags.

File: website/cue/reference/components/transforms/remap.cue [hunk 14]
```diff
@@ -8,12 +8,14 @@ components: transforms: "remap": {
...
-		the `errors` output instead. For a transform component named `foo`, this
-		error output can be accessed by specifying `foo.errors` as the input to
-		another component.
+		This transform also implements an additional `dropped` output. When the
+		`drop_on_error` or `drop_on_abort` configuration values are set to `true`,
+		events that result in runtime errors or aborts will be dropped from the
+		default output stream and sent to the `dropped` output instead. For
+		a transform component named `foo`, this output can be accessed by specifying
+		`foo.dropped` as the input to another component. Events sent to this output
+		will be in their original form, omitted any partial modification that took
```
Comment (jszwedko) [inline website/cue/reference/components/transforms/remap.cue:17]: `omitted` -> `omitting`

Comment (jszwedko) [inline website/cue/reference/components/transforms/remap.cue:17]: `omitted` -> `omitting`

File: src/transforms/remap.rs [hunk 15]
```diff
@@ -106,11 +112,45 @@ impl Remap {
...
+                            "component_type": "remap",
+                            "component_kind": "transform",
+                        }
+                    }),
+                );
+            }
+            Event::Metric(ref mut metric) => {
+                let m = log_schema().metadata_key();
+                metric.insert_tag(format!("{}.dropped.reason", m), reason.into());
+                metric.insert_tag(format!("{}.dropped.message", m), error.to_string());
```
Comment (jszwedko) [inline src/transforms/remap.rs:135]: I'd say we could probably drop message since it'd make an awkward metrics tag. We can add it later if users ask for it.

File: website/content/en/highlights/2021-11-04-remap-dropped-output.md (hunk 11 above)
Comment (jszwedko) [reply website/content/en/highlights/2021-11-04-remap-dropped-output.md:15]: This has been updated to allow metrics too.

**Review by jszwedko — COMMENTED** [review]
Just one comment about dropping the error message from tags, but otherwise looks good! I am wondering if we should have some tests though? I'd love to see some in `tests/behavior`, but at the least some component level tests around the transform would help build some confidence.

I'm also just now noticing that this does change the performance behavior. Previously, if `drop_on_error` was true and `drop_on_abort` was true then the event would not be cloned going into VRL since we don't need to preserve the original. In some ways, I like this because it makes reasoning about VRL performance a bit simpler but does remove one easy way to improve VRL perf. I'm imagining that typically users are dropping invalid events. Do you think it'd be feasible to maintain the behavior of only cloning if the original event is actually processed by checking if the error outputs stream is actually consumed by another component?

**Comment by jszwedko** [comment]
Alternatively, we let users opt into error routing through another flag like `route_error_events`? That'd probably be a simpler small fix.

**Comment by lukesteensen** [comment]
@jszwedko 

> I'd love to see some in `tests/behavior`, but at the least some component level tests around the transform would help build some confidence.

I spent a while writing a behavior test before I remembered that we explicitly chose not to integrate those with multiple outputs 😄 . I did add a pretty thorough Rust unit test though, good catch. 

> Do you think it'd be feasible to maintain the behavior of only cloning if the original event is actually processed by checking if the error outputs stream is actually consumed by another component?

Right now, no, there's not a good way to do that. At least not one that I'm willing to tackle on a PR that's already been open for almost 6 weeks 🥲 

> Alternatively, we let users opt into error routing through another flag like `route_error_events = true`? That'd probably be a simpler small fix.

Yep, that makes sense. I've added that as `reroute_dropped` since it seems like we wanted to avoid the `error` nomenclature. That makes this feature explicitly opt-in, which is probably good to start with.

**Review by jszwedko — APPROVED** [review]
🥳 

Thanks for all of the iterating on this one! I think we ended up in a good place. It'd be nice to be able to remove `reroute_dropped` in the future, but this is a great start and I'm excited to get it out in user's hands.

**Comment by jszwedko** [comment]
> I spent a while writing a behavior test before I remembered that we explicitly chose not to integrate those with multiple outputs 😄 . I did add a pretty thorough Rust unit test though, good catch.

Doh, sorry, yeah, I forgot about this too.
